src/gnutls.c

```diff
@@ -1016,82 +1017,79 @@ void cleanup_starttls (void) {
  */
 static int copycat (int local, int remote, gnutls_session_t wrapped, int client) {
 	char buf [1024];
-	struct pollfd inout [3];
 	ssize_t sz;
 	struct linger linger = { 1, 10 };
 	int have_client;
 	int retval = GNUTLS_E_SUCCESS;
 	bool local_shutdown_done = false;
 	bool remote_shutdown_done = false;
-	
-	client = -1;
-	inout [0].fd = local;
-	inout [1].fd = remote;
-#ifdef WINDOWS_PORT
-	have_client = 0;
-#else
-	inout [2].fd = client;
-	have_client = inout [2].fd >= 0;
-#endif
-	if (!have_client) {
-		inout [2].revents = 0;	// Will not be written by poll
-		//FORK!=DETACH// inout [2].fd = ctlkey_signalling_fd;
-	}
-	inout [0].events = POLLIN;
-	inout [1].events = POLLIN;
-	inout [2].events = 0;	// error events only
+	fd_set rset;
+	fd_set eset;
+	int maxfd = local > remote ? local : remote;
+	struct timeval close_timeout = { COPYCAT_CLOSE_TIMEOUT, 0 };
+
 	tlog (TLOG_COPYCAT, LOG_DEBUG, "Starting copycat cycle for local=%d, remote=%d, control=%d", local, remote, client);
-	while (((inout [0].events | inout [1].events) & POLLIN) != 0) {
-		int polled;
+	// set exception file descriptor set
+	FD_ZERO (&eset);
+	FD_SET (local, &eset);
+	FD_SET (remote, &eset);
+	if (client >= 0) {
+		FD_SET (client, &eset);
+		if (client > maxfd) {
+			maxfd = client;
+		}
+	}
+
```

Comment:
I cannot believe that this would work -- you need to `FD_ZERO()` and `FD_SET()` the `eset` as much as `rset` on every pass through the loop.  That's how `select()` works.  Maybe the `poll()` code got away with this because it set `events` to 0 for `pollfd[2]` and because it used `poll()`, but I would need to be convinced about this flavour :)